src/logic/player.cc

```diff
@@ -237,6 +241,44 @@ void Player::allocate_map() {
 	fields_.reset(new Field[map.max_index()]);
 }
 
+bool Player::pick_custom_starting_position(const Coords& c) {
+	assert(is_picking_custom_starting_position_);
+	if (!get_starting_position_suitability(c)) {
+		return false;
+	}
+	dynamic_cast<Game&>(egbase()).send_player_command(
+	   new CmdPickCustomStartingPosition(egbase().get_gametime(), player_number(), c));
+	return true;
+}
+
+void Player::do_pick_custom_starting_position(const Coords& c) {
+	if (!is_picking_custom_starting_position_) {
+		return;
+	}
+	is_picking_custom_starting_position_ = false;
+	egbase().mutable_map()->set_starting_pos(player_number(), c);
+	create_default_infrastructure();
+}
+
+bool Player::get_starting_position_suitability(const Coords& c) const {
+	const Map& map = egbase().map();
+	const FCoords f = map.get_fcoords(c);
+	if (f.field->get_owned_by() != 0 || (f.field->nodecaps() & BUILDCAPS_BIG) != BUILDCAPS_BIG) {
+		return false;
+	}
+	const Widelands::Field& neighbour = map[map.br_n(c)];
+	if (neighbour.get_owned_by() != 0 || !(neighbour.nodecaps() & BUILDCAPS_FLAG)) {
```

Comment:
I think we should check an area here too - e.g. the building's flag position might be surrounded by rocks.